### ISSUE TITLE
M4A files over 5 minutes do not play

### ISSUE THREAD

**Issue by alexfoxy** [tumtumtum/StreamingKit#310]
Hey,

I'm having an issue playing M4A files with StreakingKit. I can play M4A files under 5 minutes but any that exceed that length result in an an unexpected error (code 2) and do not play back.

I've read about M4A needing to be optimised for streaming, is this the issue with the larger files? If so how do you do so?

Thanks,
Alex

**Comment by abdultcb** [comment]
I am having the same issue but with CAF files.
I am exporting songs from iPod Library and then playing these songs.
need your help

**Comment by kuntal33** [comment]
Did you guys find any solution? I am having same issue.

**Comment by alexfoxy** [comment]
Unfortunately I just used another solution!

**Comment by kuntal33** [comment]
It would be great help if you provide me another solution that you used. Thanks in advance

**Comment by harshalrj25** [comment]
I'm having the same issue while playing M4A files with StreakingKit. I can play M4A files under 5 minutes but any that exceed that length do not play back.

**Comment by newbdez33** [comment]
Any update here?

**Comment by doriansgithub** [comment]
I believe the issue has something to do with the encoding.  If you brought in a CD with Lossless (I use XLD), then used the autoConvert to 256 in iTunes, the player won't be able to play tunes with approx. over 5:30.  I'm reloading my entire collection into iTunes at full AIFF resolution, and then using the autoConvert to 256 for the portables.  If that doesn't work, then we'll have to figure out what the issue is.

**Comment by doriansgithub** [comment]
I'm starting with my favorite band: Aerosmith, and I'm almost done, so I'll update shortly.

**Comment by doriansgithub** [comment]
Sorry to say, but that doesn't work!  I'll try encoding with AudiConverter Lite, but this is not good enough, as users will be forced to download extra software to their machine.  Most people barely know how to turn on their machines, let alone understand that there are issues with "Encoding", like what is that?

**Comment by doriansgithub** [comment]
They do play at full AIFF with no downsampling.  So the issue is reading m4a files.  I don't use mp3s, or wav so I can't speak for that.

**Comment by doriansgithub** [comment]
It seems to work with AudioConverterPro but's $6.99 and i don't want to put my customers thru that, so I hope there is another way to fix this.

**Comment by iDevelopper** [comment]
Why don't you use the AAC encoder directly when you import your CD?

**Comment by doriansgithub** [comment]
I'll try that, but ideally, I like to have the pure cd quality in iTunes so that I can hear that if wanted to.

**Comment by iDevelopper** [comment]
I realize that when I synchronise my iPhone with iTunes, I have this message when iTunes try to copy your file to the device:

![2017-06-23_12-01-22](https://user-images.githubusercontent.com/1782450/27477512-cdc2f6c6-580c-11e7-82eb-5cdd81aeb61d.png)

**Comment by doriansgithub** [comment]
It might be a lossless file.  Try converting to aac 256.

**Comment by iDevelopper** [comment]
Yes but what is your issue exactly? What is the audio file you try to play with streamingKit?

**Comment by doriansgithub** [comment]
Every file in my phone over 5:30 mins won't play.  It skips over it.

**Comment by doriansgithub** [comment]
Sorry, unless it's uncompressed (aiff), then it plays any length.

**Comment by iDevelopper** [comment]
I understand but what do you want exactly to work?

**Comment by doriansgithub** [comment]
It should play m4a files of any size that were compressed by iTunes.  I haven't tried compressing the files when bringing them in to iTunes.  But I rather have aiff files in iTunes and compressing them only when syncing to the portables.

**Comment by iDevelopper** [comment]
But lossless is not compressed!

**Comment by doriansgithub** [comment]
I believe Apple doesn't allow lossless on their portables.  I am able to load lossless in iTunes, but to load it into a device it needs to be aiff, MP3 or m4a.

**Comment by mitchdowney** [comment]
I also cannot play any M4A audio files :(

**Comment by doriansgithub** [comment]
And we still can't play m4a files that are greater than 5:30 in length, unless it's compressed by a 3rd party.

**Comment by doriansgithub** [comment]
You should be able to play m4a.  What length is the file?

**Comment by mitchdowney** [comment]
The file is over an hour and unoptimized. Maybe remote streaming playback of the file is failing because of [this old issue](https://github.com/tumtumtum/StreamingKit/issues/84#issuecomment-35248019)?

Here is the m4a file that is failing to load:

http://media.blubrry.com/jordanbpeterson/content.blubrry.com/jordanbpeterson/20_-_Ideology_Logos_and_Belief.m4a

**Comment by patrickjquinn** [comment]
It’s a 2 year old issue and Im still struggling with basically this problem, perhaps it’s time we took it into our own hands and figured out a better way to handle remote m4a streams (Or m4a at all) given that m4a has officially replaced mp3 as the dominant audio format (or so I’ve read)

**Comment by doriansgithub** [comment]
Yes, but how?  I'm not good enough to tackle low level code.  I can't see this being that difficult for Tum, since he was able to accomplish the entire project.  I think he should set up a donation format, so that he gets motivated.

**Comment by patrickjquinn** [comment]
Like any good community effort we dig into the code (its just Objective-C,
how scary can it be) and report our findings and then formulate a plan of
attack re: fixing it. From what I gather there is an AVExportSession run
over the stream as it enters the buffer (something to that effect) and i'd
imagine it relies on the header to re-encode the stream. That is where the
problem probably lies.
<https://mailtrack.io/> Sent with Mailtrack
<https://mailtrack.io/install?source=signature&lang=en&referral=patrick.jm.quinn@gmail.com&idSignature=22>

**Comment by mitchdowney** [comment]
@patrickjquinn I would love to help resolve this but unfortunately I have only Swift experience (~2 years). I'm motivated to help but would need some direction on how I could be useful.

I'd be willing to donate to get this issue resolved, if that would help.

Unfortunately @tumtumtum I like others will need to switch to FreeStreamer if we can't resolve this.

---

FWIW these open issues appear to be duplicates to me:

#84 
#89 
#171
#240 

If we should close or combine them somehow let me know.

**Comment by doriansgithub** [comment]
What is optimized?

**Comment by mitchdowney** [comment]
When I run afinfo on the file from sample link above, I see the following output. Note 6 lines from the bottom where it says "not optimized":

```
File:           /Users/asdf/desktop/24_-_The_Psychology_of_the_Flood.m4a
File type ID:   m4af
Num Tracks:     1
----
Data format:     2 ch,  44100 Hz, 'aac ' (0x00000000) 0 bits/channel, 0 bytes/packet, 1024 frames/packet, 0 bytes/frame
Channel layout: Stereo (L R)
estimated duration: 9423.624989 sec
audio bytes: 150778818
audio packets: 405844
bit rate: 127999 bits per second
packet size upper bound: 727
maximum packet size: 727
audio data file offset: 57344
not optimized
audio 415581862 valid frames + 2112 priming + 282 remainder = 415584256
format list:
[ 0] format:	  2 ch,  44100 Hz, 'aac ' (0x00000000) 0 bits/channel, 0 bytes/packet, 1024 frames/packet, 0 bytes/frame
Channel layout: Stereo (L R)
----
```

I wonder if this m4a file were optimized if it would load and seek correctly? Or if the issue affects all m4a files over 5:30 long regardless of optimization?

---

I haven't heard of m4a optimization before now, but here's someone's answer from a [message board]():

> It removes all free atoms and puts moov and tags before actual data.

> It'll make the files smaller if there is some unused space. There might be also some programs which won't read the tag if it's located after the actual data.

> – Florian

That's all I know right now. For me I don't have the rights to optimize the files, so I'll need some kind of ability to stream non-optimized m4a files with StreamingKit.

**Comment by doriansgithub** [comment]
I searched and found the same info.  Other non-streaming libraries like AudioKit and Novocaine, read these files without issue.  So it has to be something to do with the streaming portion of the player.  We are going to implement Novocaine for the local files and use strreamkit for streaming.  We obviously would rather just have streamkit, but not being able to play songs from iTunes is a game killer.

**Comment by doriansgithub** [comment]
Just remembered, that as per patrickjquinn, I used Audio Converter Lite, and the player plays the resulting m4a songs over 5:30, so maybe we can look into the difference between the encoding between iTunes and this converter.  Also maybe look at how novocaine is treating these files.

**Comment by mitchdowney** [comment]
Does anyone have a link to an m4a file over 5:30 long that StreamingKit can stream successfully?

**Comment by iDevelopper** [comment]
Sorry but it is me that recommended using Audio Converter Light.

Never mind!

Does anyone have a link to an m4a file LOWER 5:30 long that StreamingKit can stream successfully?

**Comment by iDevelopper** [comment]
Because StreamingKit works well with some m4a when there are in local but the same file from a server don't work! It is not relative to the size or longer!

**Comment by patrickjquinn** [comment]
No it’s nothing to do with the size of the file, it’s down to it’s encoding, what’s in the header and how the AVAudioSesson processes it in real time. If you can get your hand on one of the YouTube m4a files you can breakpoint this issue.

**Comment by doriansgithub** [comment]
Oops, yes, you are right, it was you that recommended AudioConverter Light, I confused patrick with you.  :(

Patrick, if the size of the file has nothing to do with it, why does it play anything under 5:30?

**Comment by iDevelopper** [comment]
Hi @doriansgithub ,

Can you give me a link to a file that works, please? Because when I put some files to dropbox, they don't play but the same files play locally!

**Comment by doriansgithub** [comment]
I can't get anything to play that is over 5:31, even after using Audio Converter Lite.  The only way for me to play anything over 5:31, is in the original AIFF format.  This works even if the file was originally a LossLess, converted to AIFF using Audio Converter Lite.

**Comment by doriansgithub** [comment]
This is what I get, I'm sure everyone else is getting the same, but I'll post it anyways.

2017-08-26 10:20:38.442040-0700 Musically[918:193200] didFinishBuffering: ipod-library://item/item.m4a?id=422125328949642500
2017-08-26 10:20:38.482391-0700 Musically[918:193200] Requeuing: ipod-library://item/item.m4a?id=422125328949642500
2017-08-26 10:20:39.416581-0700 Musically[918:193466] setCurrentlyReadingEntry:andStartPlaying:clearQueue: ipod-library://item/item.m4a?id=422125328949642500
2017-08-26 10:20:39.446686-0700 Musically[918:193831] processFinishPlayingIfAnyAndPlayingNext:withNext: Finished: ipod-library://item/item.m4a?id=422125328949642500, Next: ipod-library://item/item.m4a?id=422125328949642500, buffering.count=0,upcoming.count=0
2017-08-26 10:20:39.457476-0700 Musically[918:193200] State = 5
2017-08-26 10:20:39.457699-0700 Musically[918:193200] Finished: QueueID
2017-08-26 10:20:39.512627-0700 Musically[918:193200] Started: QueueID
2017-08-26 10:20:39.515017-0700 Musically[918:193200] State = 32
2017-08-26 10:20:39.516207-0700 Musically[918:193200] Error = 4
2017-08-26 10:20:39.949266-0700 Musically[918:193200] Error = 4
2017-08-26 10:20:39.949514-0700 Musically[918:193200] Error = 4

**Comment by iDevelopper** [comment]
You can't play some item from iPod Library directly. You have to use AVAssetExportSession first.

**Comment by iDevelopper** [comment]
@doriansgithub ,

I don't understand what you want to do exactly. Play some audios from remote url, local url or from iPod Library.

Try this sample (Please open the xcworkspace and not the xcodeproj file):

[STKSample.zip](https://github.com/tumtumtum/StreamingKit/files/1254637/STKSample.zip)

I have all the cases in it (local, remote, iTunes Library), with files (mp3, m4a) over 5:30 min or lower, purchased on iTunes or imported from a CD.

All is right...

For this sample, I did not use Audio Converter Lite.

Hope it helps!

**Comment by doriansgithub** [comment]
Ok, I'll check that out, and so should everyone else, since they are all having the same issue.  This might then might be a non-issue and just an error on everyones implementation.  I'll get back when we get this going.   Thanks Patrick!!

**Comment by doriansgithub** [comment]
I implemented AVAssetExportSession and then play exported file. I have this error: STKAudioPlayerErrorStreamParseBytesFailed. It's normally export to Caches folder on Phone, but it still doesn't play(

**Comment by doriansgithub** [comment]
I added this song to the Sample Project and it will not play.
![simulator screen shot sep 14 2017 1 15 29 pm](https://user-images.githubusercontent.com/5715600/30453787-e5280418-994e-11e7-8d44-ab2ab2383fa6.png)

Here is the file:

https://www.sendspace.com/file/gtc8z5

**Comment by iDevelopper** [comment]
It is not a duration problem as in the sample there are some songs over 5:30 who plays!

**Comment by doriansgithub** [comment]
Every 5:31+ song that I try to play fails.  I have many on my phone, and my partner has the same issue.  I'm not sure how you encoded those songs, but all I do is encode my cds into aiff or lossless, then have iTunes encode to 256.

**Comment by doriansgithub** [comment]
Also, the same songs play fine in Apple's music player, or in my app using AVAudioPlayer or using Novocaine.

**Comment by iDevelopper** [comment]
Money (9:52) is from an album I purchased in iTunes ( https://itunes.apple.com/fr/album/delicate-sound-of-thunder-live/id696774967 ). Sous Le Soleil (7:02) is from a CD I imported in iTunes with the AAC encoder.

**Comment by doriansgithub** [comment]
So both are 256.  Can you try playing the files I sent you?  Thanks!!

**Comment by doriansgithub** [comment]
The file I sent you is lossless.  I'll test an m4a.

**Comment by iDevelopper** [comment]
The file you sent cannot be play !

 

Patrick BODET

 

De : Dorian Mattar <notifications@github.com>
Répondre à : tumtumtum/StreamingKit <reply@reply.github.com>
Date : vendredi 15 septembre 2017 à 17:16
À : tumtumtum/StreamingKit <StreamingKit@noreply.github.com>
Cc : Courriel BODET <patrick.bodet4@wanadoo.fr>, Comment <comment@noreply.github.com>
Objet : Re: [tumtumtum/StreamingKit] M4A files over 5 minutes do not play (#310)

 

So both are 256. Can you try playing the files I sent you? Thanks!!

**Comment by iDevelopper** [comment]
@tumtumtum , help please on this issue.

**Comment by doriansgithub** [comment]
Hi, do you have a new sample that is iOS 11 ready, where you moved the STKPlayer to a singleton?

**Comment by andvon** [comment]
@tumtumtum , help please on this issue.

**Comment by doriansgithub** [comment]
[STKSample.zip](https://github.com/tumtumtum/StreamingKit/files/1639646/STKSample.zip)

This can play any file.

**Comment by veila** [comment]
Hey,

I'm having same issue. It works perfectly with files 320kps, but flac or m4a over 3mins, the duration always wrong ( a song 3mins but StreakingKit detect duration over 5mins, 6mins). So when finish song, the player freeze, can not play anymore and I need restart.

Did you guys find any solution?
Thanks,

**Comment by kutakmir** [comment]
Try this:
https://firebasestorage.googleapis.com/v0/b/fir-chat-dab84.appspot.com/o/test_music_dont_delete_yet%2FSTKSample.zip?alt=media&token=0cdb2ad6-e42b-41a3-a062-ce10ea8a7a8c

If the song gets over 10 minutes when playing from the start without skipping, it goes silent.
